File: app/src/main/java/com/productivity/habits/ui/detail/CircularFocusTimer.kt
package com.productivity.habits.ui.detail

import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.provider.Settings
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CenterFocusStrong
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.NotificationsOff
import androidx.compose.material.icons.filled.Pause
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.productivity.habits.data.local.preferences.ThemePreferences
import com.productivity.habits.service.FocusTimerService
import com.productivity.habits.service.TimerStateHolder
import com.productivity.habits.service.TimerStatus
import com.productivity.habits.ui.common.HapticsHelper
import java.util.Locale

@Composable
fun CircularFocusTimer(
    habitId: String,
    habitTitle: String,
    defaultDurationMinutes: Double,
    remainingUnloggedMinutes: Double = defaultDurationMinutes,
    accentColor: Color,
    themePreferences: ThemePreferences,
    onFocusScreenClick: () -> Unit = {},
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current
    val haptic = LocalHapticFeedback.current
    val timerState by TimerStateHolder.timerState.collectAsState()

    var showEditMinutesDialog by remember { mutableStateOf(false) }

    val isRunningOrPausedForThisHabit = timerState.habitId == habitId && (timerState.status == TimerStatus.RUNNING || timerState.status == TimerStatus.PAUSED)
    val isCompletedForThisHabit = timerState.habitId == habitId && timerState.status == TimerStatus.COMPLETED

    val defaultDurationSec = (defaultDurationMinutes * 60).toLong().coerceAtLeast(60L)

    val remainingSec = when {
        isRunningOrPausedForThisHabit -> timerState.remainingSeconds
        isCompletedForThisHabit -> 0L
        timerState.habitId == habitId && timerState.status == TimerStatus.IDLE -> timerState.remainingSeconds
        else -> defaultDurationSec
    }

    val totalSec = when {
        isRunningOrPausedForThisHabit -> timerState.totalSeconds
        isCompletedForThisHabit -> timerState.totalSeconds
        timerState.habitId == habitId && timerState.status == TimerStatus.IDLE -> timerState.totalSeconds
        else -> defaultDurationSec
    }

    val status = when {
        isRunningOrPausedForThisHabit -> timerState.status
        isCompletedForThisHabit -> TimerStatus.COMPLETED
        else -> TimerStatus.IDLE
    }

    val progress = if (totalSec > 0) {
        ((totalSec - remainingSec).toFloat() / totalSec.toFloat()).coerceIn(0f, 1f)
    } else 0f
    val animatedProgress by animateFloatAsState(targetValue = progress, label = "SweepTimerProgress")

    val minutes = remainingSec / 60
    val seconds = remainingSec % 60
    val timeFormatted = String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)

    if (showEditMinutesDialog) {
        var inputMins by remember { mutableStateOf((totalSec / 60).toString()) }
        AlertDialog(
            onDismissRequest = { showEditMinutesDialog = false },
            title = { Text("Set Timer Minutes", style = MaterialTheme.typography.titleMedium) },
            text = {
                OutlinedTextField(
                    value = inputMins,
                    onValueChange = { if (it.isEmpty() || it.all { c -> c.isDigit() }) inputMins = it },
                    label = { Text("Minutes") },
                    singleLine = true
                )
            },
            confirmButton = {
                TextButton(
                    onClick = {
                        val parsed = inputMins.toLongOrNull() ?: 25L
                        if (isRunningOrPausedForThisHabit) {
                            TimerStateHolder.setRemainingMinutes(parsed)
                        } else {
                            TimerStateHolder.setDuration(habitId, habitTitle, parsed.toDouble())
                        }
                        showEditMinutesDialog = false
                    }
                ) {
                    Text("Save", fontWeight = FontWeight.Bold)
                }
            },
            dismissButton = {
                TextButton(onClick = { showEditMinutesDialog = false }) {
                    Text("Cancel")
                }
            }
        )
    }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        // Top Header
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "Focus Timer",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurface
            )

            Row(
                horizontalArrangement = Arrangement.spacedBy(4.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                // Focus Screen toggle (enabled when timer is active)
                IconButton(
                    onClick = {
                        HapticsHelper.performLightHaptic(haptic)
                        onFocusScreenClick()
                    },
                    enabled = isRunningOrPausedForThisHabit,
                    modifier = Modifier.size(32.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.CenterFocusStrong,
                        contentDescription = "Enter focus mode",
                        tint = if (isRunningOrPausedForThisHabit) {
                            accentColor
                        } else {
                            MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.4f)
                        },
                        modifier = Modifier.size(20.dp)
                    )
                }

                IconButton(
                    onClick = {
                        HapticsHelper.performLightHaptic(haptic)
                        showEditMinutesDialog = true
                    },
                    modifier = Modifier.size(32.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.Edit,
                        contentDescription = "Edit timer duration",
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.size(18.dp)
                    )
                }
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Large Circular Countdown Canvas
        Box(
            modifier = Modifier.size(200.dp),
            contentAlignment = Alignment.Center
        ) {
            val trackColor = MaterialTheme.colorScheme.surfaceVariant
            val sweepAngle = animatedProgress * 360f

            Canvas(modifier = Modifier.size(190.dp)) {
                val strokeWidth = 10.dp.toPx()

                // Background track circle
                drawCircle(
                    color = trackColor,
                    style = Stroke(width = strokeWidth)
                )

                // Foreground sweep arc
                drawArc(
                    color = accentColor,
                    startAngle = -90f,
                    sweepAngle = sweepAngle,
                    useCenter = false,
                    style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
                )
            }

            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(
                    text = timeFormatted,
                    style = MaterialTheme.typography.displayMedium,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSurface,
                    letterSpacing = (-0.5).sp
                )

                val statusLabel = when (status) {
                    TimerStatus.RUNNING -> "Focusing..."
                    TimerStatus.PAUSED -> "Paused"
                    TimerStatus.COMPLETED -> "Completed!"
                    TimerStatus.IDLE -> "Ready"
                }

                Text(
                    text = statusLabel,
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = if (status == TimerStatus.RUNNING) accentColor else MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }

        Spacer(modifier = Modifier.height(20.dp))

        // Primary Control Buttons: Play/Pause Pill and Reset
        Row(
            horizontalArrangement = Arrangement.spacedBy(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            // Reset Button
            Surface(
                modifier = Modifier
                    .size(48.dp)
                    .clip(CircleShape)
                    .clickable {
                        HapticsHelper.performLightHaptic(haptic)
                        FocusTimerService.resetTimer(context)
                        TimerStateHolder.reset()
                    },
                shape = CircleShape,
                color = MaterialTheme.colorScheme.surfaceVariant
            ) {
                Box(contentAlignment = Alignment.Center) {
                    Icon(
                        imageVector = Icons.Default.Refresh,
                        contentDescription = "Reset Timer",
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.size(22.dp)
                    )
                }
            }

            // Primary Play / Pause Pill Button
            Button(
                onClick = {
                    HapticsHelper.performLightHaptic(haptic)
                    when (status) {
                        TimerStatus.RUNNING -> {
                            FocusTimerService.pauseTimer(context)
                        }
                        TimerStatus.PAUSED -> {
                            FocusTimerService.resumeTimer(context)
                        }
                        TimerStatus.COMPLETED, TimerStatus.IDLE -> {
                            val durationMins = if (totalSec > 0) totalSec / 60.0 else defaultDurationMinutes
                            FocusTimerService.startTimer(context, habitId, habitTitle, durationMins)
                        }
                    }
                },
                modifier = Modifier
                    .height(48.dp)
                    .width(160.dp),
                shape = RoundedCornerShape(24.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = accentColor,
                    contentColor = Color.White
                )
            ) {
                Icon(
                    imageVector = if (status == TimerStatus.RUNNING) Icons.Default.Pause else Icons.Default.PlayArrow,
                    contentDescription = if (status == TimerStatus.RUNNING) "Pause" else "Start",
                    modifier = Modifier.size(20.dp)
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = if (status == TimerStatus.RUNNING) "Pause" else "Start",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold
                )
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Duration Adjustment Quick Chips: -10m, -5m, [Goal], +5m, +10m
        val currentGoalMins = (totalSec / 60).toInt()
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically
        ) {
            val deltas = listOf(-10, -5, 0, 5, 10)
            deltas.forEach { delta ->
                val label = if (delta == 0) "${currentGoalMins}m Goal" else if (delta > 0) "+${delta}m" else "${delta}m"
                val isGoal = delta == 0

                FilterChip(
                    selected = isGoal,
                    onClick = {
                        if (delta != 0) {
                            HapticsHelper.performLightHaptic(haptic)
                            val newMins = maxOf(1L, (totalSec / 60) + delta)
                            if (isRunningOrPausedForThisHabit) {
                                TimerStateHolder.setRemainingMinutes(newMins)
                            } else {
                                TimerStateHolder.setDuration(habitId, habitTitle, newMins.toDouble())
                            }
                        }
                    },
                    label = {
                        Text(
                            text = label,
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = if (isGoal) FontWeight.Bold else FontWeight.Medium
                        )
                    },
                    colors = FilterChipDefaults.filterChipColors(
                        selectedContainerColor = accentColor.copy(alpha = 0.2f),
                        selectedLabelColor = accentColor
                    ),
                    modifier = Modifier.padding(horizontal = 4.dp)
                )
            }
        }

        Spacer(modifier = Modifier.height(12.dp))

        // DND Toggle Chip
        val dndEnabled by themePreferences.focusDndEnabled.collectAsState()
        val notificationManager = remember {
            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        }

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.Center
        ) {
            FilterChip(
                selected = dndEnabled,
                onClick = {
                    HapticsHelper.performLightHaptic(haptic)
                    if (!dndEnabled && !notificationManager.isNotificationPolicyAccessGranted) {
                        // Redirect to system settings to grant DND access
                        val intent = Intent(Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS)
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        context.startActivity(intent)
                    } else {
                        themePreferences.setFocusDndEnabled(!dndEnabled)
                    }
                },
                label = {
                    Text(
                        text = if (dndEnabled) "DND On" else "DND Off",
                        style = MaterialTheme.typography.labelSmall,
                        fontWeight = FontWeight.Medium
                    )
                },
                leadingIcon = {
                    Icon(
                        imageVector = if (dndEnabled) Icons.Default.NotificationsOff else Icons.Default.Notifications,
                        contentDescription = if (dndEnabled) "Disable Do Not Disturb" else "Enable Do Not Disturb",
                        modifier = Modifier.size(16.dp)
                    )
                },
                colors = FilterChipDefaults.filterChipColors(
                    selectedContainerColor = accentColor.copy(alpha = 0.2f),
                    selectedLabelColor = accentColor,
                    selectedLeadingIconColor = accentColor
                ),
                modifier = Modifier.padding(horizontal = 4.dp)
            )
        }
    }
}

File: lib/ui/detail/widgets/circular_focus_timer.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../data/preferences/theme_preferences.dart';
import '../../../services/dnd_service.dart';
import '../../common/haptics_helper.dart';
import '../controllers/timer_state_holder.dart';

class CircularFocusTimer extends ConsumerStatefulWidget {
  final String habitId;
  final String habitTitle;
  final double defaultDurationMinutes;
  final double remainingUnloggedMinutes;
  final Color accentColor;
  final VoidCallback onFocusScreenClick;

  const CircularFocusTimer({
    super.key,
    required this.habitId,
    required this.habitTitle,
    required this.defaultDurationMinutes,
    this.remainingUnloggedMinutes = 25.0,
    required this.accentColor,
    required this.onFocusScreenClick,
  });

  @override
  ConsumerState<CircularFocusTimer> createState() => _CircularFocusTimerState();
}

class _CircularFocusTimerState extends ConsumerState<CircularFocusTimer> {
  double? _customDurationMinutes;

  @override
  void didUpdateWidget(covariant CircularFocusTimer oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.remainingUnloggedMinutes != widget.remainingUnloggedMinutes ||
        oldWidget.habitId != widget.habitId ||
        oldWidget.defaultDurationMinutes != widget.defaultDurationMinutes) {
      _customDurationMinutes = null;
    }
  }

  void _showEditMinutesDialog(int currentMinutes, bool isRunningOrPaused) {
    final textController =
        TextEditingController(text: currentMinutes.toString());

    showDialog(
      context: context,
      builder: (dialogContext) {
        return AlertDialog(
          title: const Text(
            'Set Timer Minutes',
            style: TextStyle(fontWeight: FontWeight.bold),
          ),
          content: TextField(
            controller: textController,
            keyboardType: TextInputType.number,
            autofocus: true,
            decoration: InputDecoration(
              labelText: 'Minutes',
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(12),
              ),
            ),
          ),
          actions: [
            TextButton(
              onPressed: () => Navigator.of(dialogContext).pop(),
              child: const Text('Cancel'),
            ),
            TextButton(
              onPressed: () {
                final parsed = int.tryParse(textController.text) ?? 25;
                final timerNotifier =
                    ref.read(timerStateHolderProvider.notifier);
                if (isRunningOrPaused) {
                  timerNotifier.setRemainingMinutes(parsed);
                } else {
                  setState(() {
                    _customDurationMinutes = parsed.toDouble();
                  });
                  timerNotifier.setDuration(
                    widget.habitId,
                    widget.habitTitle,
                    parsed.toDouble(),
                  );
                }
                Navigator.of(dialogContext).pop();
              },
              child: const Text(
                'Save',
                style: TextStyle(fontWeight: FontWeight.bold),
              ),
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final timerState = ref.watch(timerStateHolderProvider);
    final timerNotifier = ref.read(timerStateHolderProvider.notifier);
    final dndEnabled = ref.watch(focusDndProvider);

    final isRunningOrPausedForThisHabit = timerState.habitId == widget.habitId &&
        (timerState.status == TimerStatus.running ||
            timerState.status == TimerStatus.paused);

    final isHabitCompletedToday = widget.remainingUnloggedMinutes <= 0;

    final defaultDurationSec =
        (widget.defaultDurationMinutes * 60).toInt().clamp(60, 24 * 3600);

    final remainingMinutesToUse =
        _customDurationMinutes ?? widget.remainingUnloggedMinutes;
    final remainingUnloggedSec =
        (remainingMinutesToUse * 60).toInt().clamp(0, 24 * 3600);

    final remainingSec = isRunningOrPausedForThisHabit
        ? timerState.remainingSeconds
        : (isHabitCompletedToday && _customDurationMinutes == null)
            ? 0
            : remainingUnloggedSec;

    final totalSec = isRunningOrPausedForThisHabit
        ? timerState.totalSeconds
        : _customDurationMinutes != null
            ? (_customDurationMinutes! * 60).toInt().clamp(60, 24 * 3600)
            : defaultDurationSec;

    final status = isRunningOrPausedForThisHabit
        ? timerState.status
        : (isHabitCompletedToday && _customDurationMinutes == null)
            ? TimerStatus.completed
            : TimerStatus.idle;

    final progress = totalSec > 0
        ? ((totalSec - remainingSec) / totalSec).clamp(0.0, 1.0)
        : (status == TimerStatus.completed ? 1.0 : 0.0);

    final minutes = remainingSec ~/ 60;
    final seconds = remainingSec % 60;
    final timeFormatted =
        '${minutes.toString().padLeft(2, '0')}:${seconds.toString().padLeft(2, '0')}';

    final statusLabel = switch (status) {
      TimerStatus.running => 'Focusing...',
      TimerStatus.paused => 'Paused',
      TimerStatus.completed => 'Completed!',
      TimerStatus.idle => 'Ready',
    };

    final currentDisplayMins =
        remainingSec > 0 ? (remainingSec / 60).round() : widget.defaultDurationMinutes.round();
    final goalMins = widget.defaultDurationMinutes.round();

    return Padding(
      padding: const EdgeInsets.all(16),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          // Header Row
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(
                'Focus Timer',
                style: theme.textTheme.titleMedium?.copyWith(
                  fontWeight: FontWeight.bold,
                  color: theme.colorScheme.onSurface,
                ),
              ),
              Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  // Enter Fullscreen Focus Screen
                  IconButton(
                    iconSize: 20,
                    constraints: const BoxConstraints(
                      minWidth: 32,
                      minHeight: 32,
                    ),
                    padding: EdgeInsets.zero,
                    icon: Icon(
                      Icons.center_focus_strong,
                      color: isRunningOrPausedForThisHabit
                          ? widget.accentColor
                          : theme.colorScheme.onSurfaceVariant
                              .withValues(alpha: 0.4),
                    ),
                    tooltip: 'Enter focus mode',
                    onPressed: isRunningOrPausedForThisHabit
                        ? () {
                            HapticsHelper.performLightHaptic();
                            widget.onFocusScreenClick();
                          }
                        : null,
                  ),
                  const SizedBox(width: 4),
                  // Edit Timer Duration
                  IconButton(
                    iconSize: 18,
                    constraints: const BoxConstraints(
                      minWidth: 32,
                      minHeight: 32,
                    ),
                    padding: EdgeInsets.zero,
                    icon: Icon(
                      Icons.edit,
                      color: theme.colorScheme.onSurfaceVariant,
                    ),
                    tooltip: 'Edit timer duration',
                    onPressed: () {
                      HapticsHelper.performLightHaptic();
                      _showEditMinutesDialog(
                        currentDisplayMins,
                        isRunningOrPausedForThisHabit,
                      );
                    },
                  ),
                ],
              ),
            ],
          ),

          const SizedBox(height: 16),

          // Large Circular Countdown Canvas
          SizedBox(
            width: 200,
            height: 200,
            child: Stack(
              alignment: Alignment.center,
              children: [
                CustomPaint(
                  size: const Size(190, 190),
                  painter: _FocusTimerPainter(
                    progress: progress,
                    trackColor: theme.colorScheme.surfaceContainerHighest,
                    accentColor: widget.accentColor,
                    strokeWidth: 10,
                  ),
                ),
                Column(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      timeFormatted,
                      style: theme.textTheme.displayMedium?.copyWith(
                        fontWeight: FontWeight.bold,
                        color: theme.colorScheme.onSurface,
                        letterSpacing: -0.5,
                      ),
                    ),
                    Text(
                      statusLabel,
                      style: theme.textTheme.labelMedium?.copyWith(
                        fontWeight: FontWeight.w600,
                        color: status == TimerStatus.running
                            ? widget.accentColor
                            : theme.colorScheme.onSurfaceVariant,
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ),

          const SizedBox(height: 20),

          // Primary Controls: Reset & Play/Pause Pill
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              // Reset Button (48x48)
              Material(
                color: theme.colorScheme.surfaceContainerHighest,
                shape: const CircleBorder(),
                child: InkWell(
                  customBorder: const CircleBorder(),
                  onTap: () {
                    HapticsHelper.performLightHaptic();
                    setState(() {
                      _customDurationMinutes = null;
                    });
                    timerNotifier.reset();
                  },
                  child: const SizedBox(
                    width: 48,
                    height: 48,
                    child: Icon(Icons.refresh, size: 22),
                  ),
                ),
              ),
              const SizedBox(width: 16),

              // Play / Pause Pill Button (48h x 160w)
              SizedBox(
                width: 160,
                height: 48,
                child: FilledButton.icon(
                  style: FilledButton.styleFrom(
                    backgroundColor: widget.accentColor,
                    foregroundColor: Colors.white,
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(24),
                    ),
                  ),
                  icon: Icon(
                    status == TimerStatus.running
                        ? Icons.pause
                        : Icons.play_arrow,
                    size: 20,
                  ),
                  label: Text(
                    status == TimerStatus.running
                        ? 'Pause'
                        : (status == TimerStatus.paused ? 'Resume' : 'Start'),
                    style: theme.textTheme.titleMedium?.copyWith(
                      fontWeight: FontWeight.bold,
                      color: Colors.white,
                    ),
                  ),
                  onPressed: () {
                    HapticsHelper.performLightHaptic();
                    switch (status) {
                      case TimerStatus.running:
                        timerNotifier.pause();
                        break;
                      case TimerStatus.paused:
                        timerNotifier.resume();
                        break;
                      case TimerStatus.completed:
                      case TimerStatus.idle:
                        final durationMins = remainingSec > 0
                            ? remainingSec / 60.0
                            : widget.defaultDurationMinutes;
                        timerNotifier.start(
                          widget.habitId,
                          widget.habitTitle,
                          durationMins,
                        );
                        break;
                    }
                  },
                ),
              ),
            ],
          ),

          const SizedBox(height: 16),

          // Quick Adjustment Chips: -10m, -5m, [Remaining], [Goal], +5m, +10m
          SingleChildScrollView(
            scrollDirection: Axis.horizontal,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                // -10m
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 4),
                  child: FilterChip(
                    selected: false,
                    label: const Text('-10 m'),
                    onSelected: (_) {
                      HapticsHelper.performLightHaptic();
                      final currentMins = remainingSec > 0
                          ? (remainingSec / 60).round()
                          : widget.defaultDurationMinutes.round();
                      final newMins = max(1, currentMins - 10);
                      if (isRunningOrPausedForThisHabit) {
                        timerNotifier.setRemainingMinutes(newMins);
                      } else {
                        setState(() {
                          _customDurationMinutes = newMins.toDouble();
                        });
                        timerNotifier.setDuration(
                          widget.habitId,
                          widget.habitTitle,
                          newMins.toDouble(),
                        );
                      }
                    },
                  ),
                ),
                // -5m
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 4),
                  child: FilterChip(
                    selected: false,
                    label: const Text('-5 m'),
                    onSelected: (_) {
                      HapticsHelper.performLightHaptic();
                      final currentMins = remainingSec > 0
                          ? (remainingSec / 60).round()
                          : widget.defaultDurationMinutes.round();
                      final newMins = max(1, currentMins - 5);
                      if (isRunningOrPausedForThisHabit) {
                        timerNotifier.setRemainingMinutes(newMins);
                      } else {
                        setState(() {
                          _customDurationMinutes = newMins.toDouble();
                        });
                        timerNotifier.setDuration(
                          widget.habitId,
                          widget.habitTitle,
                          newMins.toDouble(),
                        );
                      }
                    },
                  ),
                ),
                // Autofill Remaining Chip (shown when partial progress exists)
                if (widget.remainingUnloggedMinutes > 0 &&
                    widget.remainingUnloggedMinutes < widget.defaultDurationMinutes)
                  Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 4),
                    child: FilterChip(
                      selected: _customDurationMinutes == null &&
                          !isRunningOrPausedForThisHabit &&
                          status != TimerStatus.completed,
                      label: Text(
                        '${widget.remainingUnloggedMinutes.toInt()}m Remaining',
                        style: TextStyle(
                          fontWeight: (_customDurationMinutes == null &&
                                  !isRunningOrPausedForThisHabit &&
                                  status != TimerStatus.completed)
                              ? FontWeight.bold
                              : FontWeight.w500,
                          color: (_customDurationMinutes == null &&
                                  !isRunningOrPausedForThisHabit &&
                                  status != TimerStatus.completed)
                              ? widget.accentColor
                              : theme.colorScheme.onSurface,
                        ),
                      ),
                      onSelected: (_) {
                        HapticsHelper.performLightHaptic();
                        if (isRunningOrPausedForThisHabit) {
                          timerNotifier.setRemainingMinutes(
                            widget.remainingUnloggedMinutes.toInt(),
                          );
                        } else {
                          setState(() {
                            _customDurationMinutes = null;
                          });
                          timerNotifier.setDuration(
                            widget.habitId,
                            widget.habitTitle,
                            widget.remainingUnloggedMinutes,
                          );
                        }
                      },
                    ),
                  ),
                // Goal Chip
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 4),
                  child: FilterChip(
                    selected: (widget.remainingUnloggedMinutes >=
                                widget.defaultDurationMinutes &&
                            _customDurationMinutes == null &&
                            !isRunningOrPausedForThisHabit) ||
                        (_customDurationMinutes != null &&
                            _customDurationMinutes ==
                                widget.defaultDurationMinutes &&
                            !isRunningOrPausedForThisHabit),
                    label: Text(
                      '${goalMins}m Goal',
                      style: TextStyle(
                        fontWeight: ((widget.remainingUnloggedMinutes >=
                                        widget.defaultDurationMinutes &&
                                    _customDurationMinutes == null &&
                                    !isRunningOrPausedForThisHabit) ||
                                (_customDurationMinutes != null &&
                                    _customDurationMinutes ==
                                        widget.defaultDurationMinutes &&
                                    !isRunningOrPausedForThisHabit))
                            ? FontWeight.bold
                            : FontWeight.w500,
                        color: ((widget.remainingUnloggedMinutes >=
                                        widget.defaultDurationMinutes &&
                                    _customDurationMinutes == null &&
                                    !isRunningOrPausedForThisHabit) ||
                                (_customDurationMinutes != null &&
                                    _customDurationMinutes ==
                                        widget.defaultDurationMinutes &&
                                    !isRunningOrPausedForThisHabit))
                            ? widget.accentColor
                            : theme.colorScheme.onSurface,
                      ),
                    ),
                    onSelected: (_) {
                      HapticsHelper.performLightHaptic();
                      if (isRunningOrPausedForThisHabit) {
                        timerNotifier.setRemainingMinutes(goalMins);
                      } else {
                        setState(() {
                          _customDurationMinutes =
                              widget.defaultDurationMinutes;
                        });
                        timerNotifier.setDuration(
                          widget.habitId,
                          widget.habitTitle,
                          widget.defaultDurationMinutes,
                        );
                      }
                    },
                  ),
                ),
                // +5m
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 4),
                  child: FilterChip(
                    selected: false,
                    label: const Text('+5 m'),
                    onSelected: (_) {
                      HapticsHelper.performLightHaptic();
                      final currentMins = remainingSec > 0
                          ? (remainingSec / 60).round()
                          : widget.defaultDurationMinutes.round();
                      final newMins = max(1, currentMins + 5);
                      if (isRunningOrPausedForThisHabit) {
                        timerNotifier.setRemainingMinutes(newMins);
                      } else {
                        setState(() {
                          _customDurationMinutes = newMins.toDouble();
                        });
                        timerNotifier.setDuration(
                          widget.habitId,
                          widget.habitTitle,
                          newMins.toDouble(),
                        );
                      }
                    },
                  ),
                ),
                // +10m
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 4),
                  child: FilterChip(
                    selected: false,
                    label: const Text('+10 m'),
                    onSelected: (_) {
                      HapticsHelper.performLightHaptic();
                      final currentMins = remainingSec > 0
                          ? (remainingSec / 60).round()
                          : widget.defaultDurationMinutes.round();
                      final newMins = max(1, currentMins + 10);
                      if (isRunningOrPausedForThisHabit) {
                        timerNotifier.setRemainingMinutes(newMins);
                      } else {
                        setState(() {
                          _customDurationMinutes = newMins.toDouble();
                        });
                        timerNotifier.setDuration(
                          widget.habitId,
                          widget.habitTitle,
                          newMins.toDouble(),
                        );
                      }
                    },
                  ),
                ),
              ],
            ),
          ),

          const SizedBox(height: 12),

          // DND Toggle Chip
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              FilterChip(
                selected: dndEnabled,
                showCheckmark: false,
                avatar: Icon(
                  dndEnabled
                      ? Icons.notifications_off
                      : Icons.notifications,
                  size: 16,
                  color: dndEnabled
                      ? widget.accentColor
                      : theme.colorScheme.onSurfaceVariant,
                ),
                label: Text(
                  dndEnabled ? 'DND On' : 'DND Off',
                  style: theme.textTheme.labelSmall?.copyWith(
                    fontWeight: FontWeight.w500,
                    color: dndEnabled
                        ? widget.accentColor
                        : theme.colorScheme.onSurface,
                  ),
                ),
                selectedColor: widget.accentColor.withValues(alpha: 0.2),
                onSelected: (_) async {
                  HapticsHelper.performLightHaptic();
                  if (!dndEnabled) {
                    final granted = await DndService.isDndAccessGranted();
                    if (!granted) {
                      await DndService.openDndSettings();
                      return;
                    }
                  }
                  await ref
                      .read(focusDndProvider.notifier)
                      .setFocusDndEnabled(!dndEnabled);
                },
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class _FocusTimerPainter extends CustomPainter {
  final double progress;
  final Color trackColor;
  final Color accentColor;
  final double strokeWidth;

  _FocusTimerPainter({
    required this.progress,
    required this.trackColor,
    required this.accentColor,
    required this.strokeWidth,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final radius = (min(size.width, size.height) - strokeWidth) / 2;

    // Background track circle
    final trackPaint = Paint()
      ..color = trackColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;
    canvas.drawCircle(center, radius, trackPaint);

    // Foreground sweep arc
    if (progress > 0) {
      final progressPaint = Paint()
        ..color = accentColor
        ..style = PaintingStyle.stroke
        ..strokeCap = StrokeCap.round
        ..strokeWidth = strokeWidth;

      canvas.drawArc(
        Rect.fromCircle(center: center, radius: radius),
        -pi / 2,
        progress * 2 * pi,
        false,
        progressPaint,
      );
    }
  }

  @override
  bool shouldRepaint(covariant _FocusTimerPainter oldDelegate) {
    return oldDelegate.progress != progress ||
        oldDelegate.trackColor != trackColor ||
        oldDelegate.accentColor != accentColor ||
        oldDelegate.strokeWidth != strokeWidth;
  }
}
